### PR TITLE
#1064 add copy_paste logging and more error copy-paste error handling

### PR DESCRIPTION
**Pull request by alex-courtis**
- Added logs to copy-paste
- Added more error and sanity checks to copy-paste

The diffs shown in the PR are pretty difficult to read. Best to checkout and `git diff -w `.
do_copy - many changes
do_single_paste, do_paste - some changes
do_cut - minimal changes

### CODE REVIEW

File: lua/nvim-tree/actions/copy-paste.lua [hunk 1]
```diff
@@ -123,9 +169,17 @@ local function do_paste(node, action_type, action_fn)
 end
 
 local function do_cut(source, destination)
-  local success = uv.fs_rename(source, destination)
+  log.line("copy_paste", "do_cut '%s' -> '%s'", source, destination)
+
+  if source == destination then
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/copy-paste.lua:174]: Exiting early can prevent some issue with recursive directory copying... onto itself...

File: lua/nvim-tree/actions/copy-paste.lua [hunk 2]
```diff
@@ -63,9 +103,10 @@ local function do_single_paste(source, dest, action_type, action_fn)
   end
 
   if should_process then
-    local success, errmsg = action_fn(source, dest)
+    success, errmsg = action_fn(source, dest)
     if not success then
-      a.nvim_err_writeln("Could not " .. action_type .. " " .. source .. " - " .. errmsg)
+      a.nvim_err_writeln("Could not " .. action_type .. " " .. source .. " - " .. (errmsg or "???"))
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/copy-paste.lua:108]: We were getting a few "can't concat nil errmsg".

File: lua/nvim-tree/actions/copy-paste.lua [hunk 3]
```diff
@@ -13,39 +14,78 @@ local clipboard = {
...
+
+      local new_name = utils.path_join { source, name }
+      local new_destination = utils.path_join { destination, name }
+      success, errmsg = do_copy(new_name, new_destination)
+      if not success then
+        return false, errmsg
+      end
+    end
+  else
+    errmsg = string.format("'%s' illegal file type '%s'", source, source_stats.type)
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/copy-paste.lua:69]: trying to copy sockets fals hard

File: lua/nvim-tree/actions/copy-paste.lua [hunk 4]
```diff
@@ -13,39 +14,78 @@ local clipboard = {
...
+  if source_stats.type == "file" then
+    success, errmsg = uv.fs_copyfile(source, destination)
+    if not success then
+      log.line("copy_paste", "do_copy fs_copyfile failed '%s'", errmsg)
+      return false, errmsg
+    end
+    return true
+  elseif source_stats.type == "directory" then
+    handle, errmsg = uv.fs_scandir(source)
+    if type(handle) == "string" then
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/copy-paste.lua:42]: I cannot imagine the situation in which a string is returned. I have left it here unchanged.

**Review by kyazdani42 — APPROVED** [review]

**Pull request merged**